custom AuthConnector for OAuth2

gusega opened the pull request:
## Hey, I just made a Pull Request!

Provide custom authConnector to Oauth2

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -551,6 +573,9 @@ export class OAuth2
 export type OAuth2CreateOptions = OAuthApiCreateOptions & {
   scopeTransform?: (scopes: string[]) => string[];
   popupOptions?: PopupOptions;
+  authConnectorFactory?: (
+    opts: OAuth2CreateOptions,
+  ) => AuthConnector<OAuth2Session>;
```
Could this just be `authConnector?: AuthConnector<OAuth2Session>`? The `opts` type isn't really correct since these fields have been defaulted, but we wouldn't want to add another type. Overall it seems a bit unnecessary to wrap up like this

gusega replied:
ok, got it. The downside of providing ready `AuthConnector` is that user must take care of providing same `scopeTransform` option to both `OAuth2#create(...)` and in the passed connector.

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -478,6 +491,15 @@ export class LocalStorageFeatureFlags implements FeatureFlagsApi {
+export type LoginPopupOptions = {
+  url: string;
+  name: string;
+  origin: string;
```
Could you move this into `showLoginPopup` to be derived from the `url` instead?

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -478,6 +491,15 @@ export class LocalStorageFeatureFlags implements FeatureFlagsApi {
 }
 
+// @public
+export type LoginPopupOptions = {
```
Rename this one to `ShowLoginPopupOptions`, or matching w/e name the function ends up with

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -637,6 +662,9 @@ export class SamlAuth
 }
 
+// @public
+export function showLoginPopup(options: LoginPopupOptions): Promise<any>;
```
Thinking `openLoginPopup` might be a better name for this one?

It needs a more explicit return type as well, can't have `any`. Needs to either be narrowed down or switched to `unknown`

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -353,6 +360,12 @@ export function createFetchApi(options: {
 }): FetchApi;
 
+// @public (undocumented)
+export type CreateSessionOptions = {
```
Rename this ot `AuthConnectorCreateSessionOptions`

Rugvip commented on packages/core-app-api/report.api.md:
```diff
@@ -294,6 +294,13 @@ export type AuthApiCreateOptions = {
+// @public
+export type AuthConnector<AuthSession> = {
+  createSession(options: CreateSessionOptions): Promise<AuthSession>;
+  refreshSession(scopes?: Set<string>): Promise<AuthSession>;
```
Could you wrap this one up in options as well? So `options?: AuthConnectorRefreshSessionOptions` with `scopes` inside

Rugvip commented on .changeset/brave-pandas-beam.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/core-app-api': patch
```
Gonna be a `minor` for this one given the new API

Rugvip commented on .changeset/brave-pandas-beam.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/core-app-api': patch
+---
+
+custom AuthConnector for OAuth2
```
Could you fill this out a little bit more please?

Rugvip reviewed (changes requested):
Thank you! 👍 

The new API surface looks alright and we can go ahead with this. Just a bunch of nits now that these things around about to become public API, things we don't care about as much for internal APIs.

gusega commented:
Hey guys, thanks a lot for the comments, I will take a look at them and update as soon as I am back from vacation (jan 6). I leave this comment here so that PR won't automatically close.

gusega commented:
@Rugvip I updated the PR, please check.

freben commented:
@gusega did you mark it as draft to indicate it's still being worked on and not ready for review?

gusega commented:
@freben I want to add a unit test and slightly change the exposed interface, I am working it at the moment. I will mark request ready afterwards.

gusega commented:
@freben  I want to add a unit test. I will mark it as ready after that.

gusega commented:
@freben  
1. I added a unit test to verify that I can create an `OAuth2` with a custom auth connector.
2. fixed `OAuth2CreateOptions` export, previosly it was not exported
3. exported `OAuth2Response` 
4. made `OAuth2#normalizeScopes` public
5. changed the implementation of `OAuth2#create`, now it accepts an options that is a union of `OAuth2CreateOptions | OAuth2CreateOptionsWithAuthConnector` and creates `AuthConnector` depending on the options provided. Previously `authConnector` was an optional property on `OAuth2CreateOptions`

Please review.

Rugvip reviewed (changes requested):
That you for the updates! 👍 

Just a few more bits to iron out and then good to :shipit:

gusega commented on packages/core-app-api/report.api.md:
```diff
@@ -553,6 +583,29 @@ export type OAuth2CreateOptions = OAuthApiCreateOptions & {
+};
+
+// @public (undocumented)
+export type OAuth2Response = {
```
@Rugvip well, almost. `OAuth2Response` represents raw OAuth2 response from the backend.

Custom auth connector may call backend auth plugin. `OAuth2Response` makes sure that custom auth connector can transform it to `OAuth2Session`.

Rugvip replied:
Alright, let's skip exporting this and leave it to the custom auth connector implementation to replicate it if needed

gusega replied:
updated.

gusega commented:
bump

gusega commented on packages/core-app-api/report.api.md:
```diff
@@ -540,6 +565,11 @@ export class OAuth2
+  static normalizeScopes(
+    scopeTransform: (scopes: string[]) => string[],
+    scopes?: string | string[],
+  ): Set<string>;
```
updated.

freben commented:
We'll get back to you - we're going to kubecon and have been a bit swamped but you're not forgotten!

gusega commented:
@freben  hi, any updates?

Rugvip reviewed (approved):
Yep it all looks good now, thank you! 👍 🎉 

Let's :shipit:
